src/Tools/Pagination/Paginator.php

```diff
@@ -51,6 +75,190 @@ public function __construct(
         $this->query = $query;
     }
 
+    /**
+     * Create an instance of Paginator with auto-detection of whether the provided
+     * query is suitable for simple (and fast) pagination queries, or whether a complex
+     * set of pagination queries has to be used.
+     */
+    public static function newWithAutoDetection(Query|QueryBuilder $query): self
+    {
+        if ($query instanceof QueryBuilder) {
+            $query = $query->getQuery();
+        }
+
+        $queryAST = $query->getAST();
+        [
+            'queryHasGroupByClause' => $queryHasGroupByClause,
+            'queryHasHavingClause' => $queryHasHavingClause,
+            'rootEntityHasSingleIdentifierFieldName' => $rootEntityHasSingleIdentifierFieldName,
+            'queryCouldProduceDuplicates' => $queryCouldProduceDuplicates,
+            'queryCouldHaveToManyJoins' => $queryCouldHaveToManyJoins,
+        ]         = self::autoDetectQueryFeatures($query->getEntityManager(), $queryAST);
+
+        $paginator = new self($query, $queryCouldProduceDuplicates);
+
+        $paginator->queryStyleAutoDetectionEnabled = true;
+        $paginator->queryCouldHaveToManyJoins      = $queryCouldHaveToManyJoins;
+        // The following is ensuring the conditions for when the CountWalker cannot be used.
+        $paginator->useCountQueryOutputWalker = $queryHasHavingClause !== false
+            || $rootEntityHasSingleIdentifierFieldName !== true
+            || ($queryCouldHaveToManyJoins && $queryHasGroupByClause !== false);
+
+        return $paginator;
+    }
+
+    /**
+     * @return array{
+     *  queryHasGroupByClause: bool|null,
+     *  queryHasHavingClause: bool|null,
+     *  rootEntityHasSingleIdentifierFieldName: bool|null,
+     *  queryCouldProduceDuplicates: bool,
+     *  queryCouldHaveToManyJoins: bool,
```

Comment:
these `query` prefixes seem unnecessary.

Reply:
The prefix "query" for the array keys here is there because some of these "keys" then "become" instance variables (of the same name). But I will drop that prefix from all the keys here, as you proposed.